atl-checker/src/algorithms/certain_zero/search_strategy/linear_constrained_phi.rs

```diff
@@ -0,0 +1,253 @@
+use crate::algorithms::certain_zero::search_strategy::linear_constraints::{
+    ComparisonOp, LinearConstraint, LinearConstraintExtractor,
+};
+use crate::atl::Phi;
+use crate::game_structure::lcgs::ast::{
+    BinaryOpKind, DeclKind, Expr, ExprKind, Identifier, UnaryOpKind,
+};
+use crate::game_structure::lcgs::ir::intermediate::IntermediateLcgs;
+use crate::game_structure::Proposition;
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+/// A structure describing the relation between linear constraints in a ATL formula
+#[derive(Clone)]
+pub enum LinearConstrainedPhi {
+    /// Either or should hold
+    Or(Box<LinearConstrainedPhi>, Box<LinearConstrainedPhi>),
+    /// Both should hold
+    And(Box<LinearConstrainedPhi>, Box<LinearConstrainedPhi>),
+    /// Mapping symbols to ranges
+    Constraint(LinearConstraint),
+    /// Proposition that is not linear. We can't use these in the linear methods, but it is not
+    /// definitively true or false either
+    NonLinear,
+    True,
+    False,
+}
+
+pub struct ConstrainedPhiMapper {
+    /// Cached LinearConstrainedPhi of propositions. The bool is true for negated propositions.
+    proposition_cache: RefCell<HashMap<(Proposition, bool), LinearConstrainedPhi>>,
```

Comment:
Instead of passing Game around in the mapping function ,couldnt it just be part of the struct?

Reply:
I tried given a lifetimed reference to the game in the structure, borrow from LinearProgrammingSearch, but Rust was not happy about it. Found a forum post that said references to parent struct are generally difficult and the best alternatives are 1) cloning the thing, or 2) passing the reference on every call, or 3) use a `Rc<RefCell<T>>` (this option doesn't work for us since the search strategy has to be `'static` which Rc isn't). So I those the second option since there will be too many clones otherwise.